rules/custom-error.js

```diff
@@ -0,0 +1,140 @@
+'use strict';
+const upperfirst = require('lodash.upperfirst');
+
+const nameRegexp = /^(?:[A-Z][a-z0-9]*)*Error$/;
+
+const getClassName = name => {
+	name = upperfirst(name);
+
+	const suffix = name.slice(-5);
+
+	if (suffix === 'error') {
+		return name.slice(0, name.length - 5) + 'Error';
+	}
+
+	return suffix === 'Error' ? name : `${name}Error`;
+};
+
+const getConstructorMethod = className => `
+	constructor() {
+		super();
+		this.name = '${className}';
+	}
+`;
+
+const isSuperExpression = node => node.type === 'ExpressionStatement' && node.expression.type === 'CallExpression' && node.expression.callee.type === 'Super';
+
+const isAssignmentExpression = (node, name) => {
+	if (node.type !== 'ExpressionStatement' || node.expression.type !== 'AssignmentExpression') {
+		return false;
+	}
+
+	const lhs = node.expression.left;
+
+	if (!lhs.object || lhs.object.type !== 'ThisExpression') {
+		return false;
+	}
+
+	return lhs.property.name === name;
+};
+
+const create = context => {
+	return {
+		ClassDeclaration: node => {
+			if (!node.superClass || !nameRegexp.test(node.superClass.name)) {
+				return;
+			}
+
+			const name = node.id.name;
+			const className = getClassName(name);
+
+			if (name !== className) {
+				context.report({
+					node: node.id,
+					message: `Invalid class name, use \`${className}\`.`,
+					fix: fixer => fixer.replaceText(node.id, className)
```

Comment:
This rule isn't fixed automatically, but I assume we should still set `this.name` automatically to the correct class name?

```js
class fooError extends Error {
    constructor() {
        super();
        this.name = 'foo';
    }
}
```

would become


```js
class fooError extends Error {
    constructor() {
        super();
        this.name = 'FooError';
    }
}
```